permuta/perm_utils/insertion_encodable.py

```diff
@@ -0,0 +1,92 @@
+from itertools import islice
+from typing import ClassVar, Dict, Iterable, Tuple
+
+from permuta.patterns.perm import Perm
+
+
+class InsertionEncodablePerms:
+    """A static container of functions fortesting
+    if a basis has a regular insertion encoding.
+    """
+
+    _ALL_PROPERTIES: ClassVar[int] = 15
+    _CACHE: ClassVar[Dict[Tuple, int]] = dict()
+
+    @staticmethod
+    def _is_incr_next_incr(perm: Perm) -> bool:
+        n = len(perm)
+        return not any(
+            curr < prev and any(perm[j + 1] < perm[j] for j in range(i + 1, n - 1))
+            for i, (prev, curr) in enumerate(zip(perm, islice(perm, 1, None)))
+        )
+
+    @staticmethod
+    def _is_incr_next_decr(perm: Perm) -> bool:
+        n = len(perm)
+        return not any(
+            curr < prev and any(perm[j + 1] > perm[j] for j in range(i + 1, n - 1))
+            for i, (prev, curr) in enumerate(zip(perm, islice(perm, 1, None)))
+        )
+
+    @staticmethod
+    def _is_decr_next_incr(perm: Perm) -> bool:
+        n = len(perm)
+        return not any(
+            curr > prev and any(perm[j + 1] < perm[j] for j in range(i + 1, n - 1))
+            for i, (prev, curr) in enumerate(zip(perm, islice(perm, 1, None)))
+        )
+
+    @staticmethod
+    def _is_decr_next_decr(perm: Perm) -> bool:
+        n = len(perm)
+        return not any(
+            curr > prev and any(perm[j + 1] > perm[j] for j in range(i + 1, n - 1))
+            for i, (prev, curr) in enumerate(zip(perm, islice(perm, 1, None)))
+        )
```

Comment:
Notice these properties are the same as some of the properties in the is_polynomial functions. Is it worth taking that into consideration?

Reply:
Yes, I think so. I'll look into it in the "next batch of refactoring"